MdeModulePkg/Include/Guid/MmMemoryProtectionSettings.h

```diff
@@ -0,0 +1,223 @@
+    UINT8    EfiPalCode                 : 1;
+    UINT8    EfiPersistentMemory        : 1;
+    UINT8    EfiUnacceptedMemoryType    : 1;
+    UINT8    OEMReserved                : 1;
+    UINT8    OSReserved                 : 1;
+  } Fields;
+} MM_HEAP_GUARD_MEMORY_TYPES;
+
+typedef UINT8 MM_MEMORY_PROTECTION_SETTINGS_VERSION;
+
+#define MM_MEMORY_PROTECTION_SETTINGS_CURRENT_VERSION  2 // Current iteration of MM_MEMORY_PROTECTION_SETTINGS
+
+//
+// Memory Protection Settings struct
+//
+typedef struct {
+  // The current version of the structure definition. This is used to ensure there isn't a definition mismatch
+  // if modules have differing iterations of this header. When creating this struct, use the
+  // MM_MEMORY_PROTECTION_SETTINGS_CURRENT_VERSION macro.
+  MM_MEMORY_PROTECTION_SETTINGS_VERSION    StructVersion;
+
+  // If enabled, accessing NULL address in UEFI or SMM code can be caught by marking
+  // the NULL page as not present.
+  BOOLEAN                                  NullPointerDetectionPolicy;
+
+  // Bitfield to control Heap Guard behavior.
+  //
+  // Note:
+  //  a) Due to the limit of pool memory implementation and the alignment
+  //     requirement of UEFI spec, HeapGuardPolicy.Direction is a try-best
+  //     setting which cannot guarantee that the returned pool is exactly
+  //     adjacent to head guard page or tail guard page.
+  //
+  //  .MmPageGuard          : Enable SMM page guard.
+  //  .MmPoolGuard          : Enable SMM pool guard.
+  //  .Direction             : The direction of Guard Page for Pool Guard.
+  //                           0 - The returned pool is near the tail guard page.
+  //                           1 - The returned pool is near the head guard page.
+  MM_HEAP_GUARD_POLICY    HeapGuardPolicy;
+
+  // Indicates which type allocation need guard page.
+  //
+  // If bit is set, a head guard page and a tail guard page will be added just
+  // before and after corresponding type of pages which the allocated pool occupies,
+  // if there's enough free memory for all of them. The pool allocation for the
+  // type related to cleared bits keeps the same as ususal.
+  //
+  // This bitfield is only valid if UefiPoolGuard and/or MmPoolGuard are set in HeapGuardPolicy.
+  MM_HEAP_GUARD_MEMORY_TYPES    HeapGuardPoolType;
+
+  // Indicates which type allocation need guard page.
+  //
+  // If a bit is set, a head guard page and a tail guard page will be added just
+  // before and after corresponding type of pages allocated if there's enough
+  // free pages for all of them. The page allocation for the type related to
+  // cleared bits keeps the same as ususal.
+  //
+  // This bitfield is only valid if MmPageGuard is set in HeapGuardPolicy.
+  MM_HEAP_GUARD_MEMORY_TYPES    HeapGuardPageType;
+} MM_MEMORY_PROTECTION_SETTINGS;
+
+#define HOB_MM_MEMORY_PROTECTION_SETTINGS_GUID \
+  { \
+    { 0x0CF445DD, 0xA67C, 0x4F8C, { 0x81, 0x9B, 0xB7, 0xB6, 0x86, 0xED, 0x7C, 0x75 } } \
+  }

```

Comment:
If the macro in `DxeMemoryProtectionSettings.h` is moved, please move this one as well.